### PR TITLE
Respect hasMoreRows returned from HS2

### LINKED ISSUE #369: Impyla handles empty results with hasMoreRows=true incorrectly
I'm trying to enable Impala's tests to run with the hs2-http protocol we added in Impyla #359 and released in 0.16.0. My current work in progress is here: https://gerrit.cloudera.org/#/c/14059/

As you can see all tests passed in PS8. Since then, Impala has added support for result spooling and as part of that change it is now possible to receive empty results with no rows in them but with hasMoreRows set to true. In this case, Impyla makes the wrong decision that no more data needs to be fetched:

```
-- 2019-10-07 22:10:10,703 DEBUG    MainThread: FetchResults: req=TFetchResultsReq(operationHandle=TOperationHandle(hasResultSet=True, modifiedRowCount=None, operationType=0, operationId=THandleIdentifier(secret='\xf5W]L\xc4wM\x85\x84\xbe\xe3\xd6\x0fEC\xeb', guid=' N\xc6\xdb\x13CMv\x00\x00\x00\x00\x0cV1(')), orientation=0, maxRows=1024)
-- 2019-10-07 22:10:10,703 DEBUG    MainThread: Attempting to open transport (tries_left=3)
-- 2019-10-07 22:10:10,703 DEBUG    MainThread: Transport opened
-- 2019-10-07 22:10:10,721 DEBUG    MainThread: FetchResults: resp=TFetchResultsResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), results=TRowSet(rows=[], columns=[TColumn(i32Val=None, byteVal=None, i16Val=None, i64Val=None, stringVal=TStringColumn(nulls='', values=[]), boolVal=None, doubleVal=None, binaryVal=None), TColumn(i32Val=None, byteVal=None, i16Val=None, i64Val=None, stringVal=TStringColumn(nulls='', values=[]), boolVal=None, doubleVal=None, binaryVal=None), TColumn(i32Val=None, byteVal=None, i16Val=None, i64Val=None, stringVal=TStringColumn(nulls='', values=[]), boolVal=None, doubleVal=None, binaryVal=None)], startRowOffset=2), hasMoreRows=True)
-- 2019-10-07 22:10:10,721 DEBUG    MainThread: fetch_results: constructing CBatch
-- 2019-10-07 22:10:10,721 DEBUG    MainThread: CBatch: input TRowSet num_cols=3 num_rows=0 tcols=[TStringColumn(nulls='', values=[]), TStringColumn(nulls='', values=[]), TStringColumn(nulls='', values=[])]
-- 2019-10-07 22:10:10,721 DEBUG    MainThread: __next__: no more data to fetch
```

Impala is only supposed to return an empty result set after a configurable timeout (10min by default) and I'm still investigating why this happens after only a few ms here. Nonetheless this is a bug in Impyla and we need to fix it.

### PR DESCRIPTION
**Pull request by timarmstrong**
This fixes #369.

Testing:
Added a test that reproduces the issue. It failed before
the fix but now passes

### CODE REVIEW

File: impala/hiveserver2.py [hunk 1]
```diff
@@ -566,26 +568,29 @@ def next(self):
...
+            elif self._last_operation_active:
+                log.debug('__next__: buffer empty and op is active => fetching '
+                          'more data')
+                self._buffer = self._last_operation.fetch(self.description,
+                                                          self.buffersize,
+                                                          convert_types=self.convert_types)
+                if len(self._buffer) > 0:
+                  log.debug('__next__: popping row out of buffer')
+                  return self._buffer.pop()
+                if not self._buffer.has_more_rows:
```
Comment (lekv) [inline impala/hiveserver2.py]: nit: I wonder if this was easier to parse if we named the flag `expect_more_rows`

File: impala/hiveserver2.py [hunk 2]
```diff
@@ -481,7 +481,9 @@ def fetchone(self):
 
     def fetchcbatch(self):
         '''Return a CBatch object of any data currently in the buffer or
```
Comment (lekv) [inline impala/hiveserver2.py]: nit: not your change, but could prettify the grammar of the existing comment.

File: impala/hiveserver2.py (hunk 1 above)
Comment (timarmstrong) [reply impala/hiveserver2.py]: Done

File: impala/hiveserver2.py (hunk 2 above)
Comment (timarmstrong) [reply impala/hiveserver2.py]: Done

**Pull request merged**